source/credentials_provider_sso.c

```diff
@@ -0,0 +1,1017 @@
+        &provider,
+        sizeof(struct aws_credentials_provider),
+        &impl,
+        sizeof(struct aws_credentials_provider_sso_impl));
+    if (!provider || !impl) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*provider);
+    AWS_ZERO_STRUCT(*impl);
+
+    aws_credentials_provider_init_base(provider, allocator, &s_aws_credentials_provider_sso_vtable, impl);
+
+    aws_tls_connection_options_init_from_ctx(&tls_connection_options, options->tls_ctx);
+    /* Override the default timeout (see STS web identity provider). */
+    tls_connection_options.timeout_ms = 3 * SSO_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS * 1000 /*msec*/;
+
+    parameters = s_parameters_new(allocator);
+    if (!parameters) {
+        goto error;
+    }
+
+    struct aws_byte_cursor host = aws_byte_cursor_from_buf(&parameters->endpoint);
+    if (aws_tls_connection_options_set_server_name(&tls_connection_options, allocator, &host)) {
+        AWS_LOGF_INFO(AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "sso: failed to create TLS connection: %s", aws_error_str(aws_last_error()));
+        goto error;
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    socket_options.type = AWS_SOCKET_STREAM;
+    socket_options.domain = AWS_SOCKET_IPV4;
+    socket_options.connect_timeout_ms = (uint32_t)aws_timestamp_convert(
+        SSO_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_MILLIS, NULL);
+
+    struct aws_http_connection_manager_options manager_options;
+    AWS_ZERO_STRUCT(manager_options);
+    manager_options.bootstrap = options->bootstrap;
+    manager_options.initial_window_size = SSO_RESPONSE_SIZE_LIMIT;
+    manager_options.socket_options = &socket_options;
+    manager_options.host = host;
+    manager_options.port = 443;
+    manager_options.max_connections = 2;
+    manager_options.shutdown_complete_callback = s_on_connection_manager_shutdown;
+    manager_options.shutdown_complete_user_data = provider;
+    manager_options.tls_connection_options = &tls_connection_options;
+
+    impl->function_table = options->function_table;
+    if (impl->function_table == NULL) {
+        impl->function_table = &s_default_function_table;
+    }
+
+    impl->connection_manager = impl->function_table->aws_http_connection_manager_new(allocator, &manager_options);
+    if (impl->connection_manager == NULL) {
+        goto error;
+    }
+
+    struct aws_exponential_backoff_retry_options retry_options = {
+        .el_group = options->bootstrap->event_loop_group,
+        .max_retries = SSO_MAX_RETRIES,
+    };
+    impl->retry_strategy = aws_retry_strategy_new_exponential_backoff(allocator, &retry_options);
+    if (!impl->retry_strategy) {
+        goto error;
+    }
+
+    impl->endpoint = aws_string_new_from_buf(allocator, &parameters->endpoint);
+    if (impl->endpoint == NULL) {
+        goto error;
+    }
+
+    impl->access_token = aws_string_new_from_buf(allocator, &parameters->access_token);
```

Comment:
Ugh, I just realized after looking at the tests that we are failing provider creation on a lot of conditions that shouldn't be failures (expired token, missing token) and it's arguable that we shouldn't be failing on other situations either (profile missing, no profile file).  Instead, those should be get credentials failures.  All of this caching and calculating we're doing ahead of time should be deferred to credential query time.  If I have a long-running process, it's going to need to refresh that access token based on expiration.  And if the provider fails to create because of these conditions then even if I fix them later by refreshing or modifying the profile appropriately, it won't get used because the provider never got created.

Reply:
I am not sure about this - it looks like an architectural question.
Currently the behaviour is - configuration and valid access token missing at the begin - don't use the SSO provider.

The main SDK also does not seem to have a good answer to this question yet: if the access token is missing or expired, it will just [add empty/void credentials to each request](https://github.com/aws/aws-sdk-cpp/blob/main/aws-cpp-sdk-core/source/auth/SSOCredentialsProvider.cpp#L59-L67). 
This results in error responses like the following:
```GetObject failed: HTTP response code: -1
Resolved remote host IP address: 
Request ID: 
Exception name: 
Error message: 
0 response headers:
```
It is very difficult to relate that back to a credentials problem, which is why in our code we have replaced the second return statement with `throw std::runtime_error`.

I can see a point for "access token expired" and will look into that next week. For now updating the PR with regard to above comments.

Reply:
Well, I think a rule-of-thumb (which we may not live up to everywhere, but we don't want to make things worse) we could use is "credentials provider instantiation should not be dependent on any transient state from the process's perspective."  When part of a larger chain, it should just be a get-credentials failure that defers to the next provider in the chain.

Reply:
@bretambrose - would it be possible to narrow down which failures are considered transient? I looked into the above, and I believe it would not help (we would not be able to use such an implementation).

The reason is that if the SSO provider is the only suitable credentials provider in a given environment (for instance in a k8s cluster), the precondition for the code should be that everything that the SSO provider needs to run should be there (profile set-up, up-to-date access token). 

Trying to compensate for a missing precondition in code is complex and error-prone, because if the interaction with outside processes (e.g. a long-running process to refresh an expired access token) does not run smoothly, the call to `s_aws_provider_chain_member_callback` will move on to the next provider (`ecs_or_imds_provider`), and requests will still fail with arcane error messages (the ECS or IMDS provider either failing to provide credentials, or unsuitable credentials).

Hence, would you be ok to regard the following conditions as required preconditions for the SSO provider to work?:
- `~/.aws/config` missing or malformed,
- `~/.aws/config` not containing the specified `profile`,
- Access token in `~/.aws/sso/cache/<sha1>.json` missing,
- Access token expired (since this is outside of the control of the `aws-c-auth` code and needs to be done by an external process)?

Reply:
If you're using the default chain I don't see how the error situation is improved by skipping SSO provider creation.  In either case you'll still fully fail to source credentials.  At least if the SSO provider was present, you'd see its sourcing failure in logs.

There isn't any precedent for dynamic state causing us to drop providers entirely, failing their creation.  The closest is the conditional ecs-or-imds check.  Beyond that, the internal specification of the SSO provider implicitly precludes these failures being construction time by explicitly calling them out as potential source-time failures.

Reply:
@bretambrose - would it meet the requirements if the SSO provider logs its failure using something like?:
```
sso: unable to configure provider (reason ...)
```

Reply:
Since the main SDK provides [credential delegation](https://github.com/aws/aws-sdk-cpp/commit/de252525aa140cbc5822e4eeedb19f34deead66b), we are likely going to use that for SSO support.

Happy to correspond on the PR, though.